File: Falling Apples/apples.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// ReadInput reads the input file and returns the grid, rows, and columns
func ReadInput(filename string) ([][]rune, int, int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, 0, 0, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// Read the first line for rows and columns
	scanner.Scan()
	firstLine := scanner.Text()
	dimensions := strings.Fields(firstLine)
	rows, err := strconv.Atoi(dimensions[0])
	if err != nil {
		return nil, 0, 0, err
	}
	cols, err := strconv.Atoi(dimensions[1])
	if err != nil {
		return nil, 0, 0, err
	}

	// Read the grid
	grid := make([][]rune, rows)
	for i := 0; i < rows; i++ {
		scanner.Scan()
		grid[i] = []rune(scanner.Text())
	}

	return grid, rows, cols, nil
}

// SimulateFallingApples simulates the apples falling in the grid
func SimulateFallingApples(grid [][]rune, rows, cols int) [][]rune {
	for col := 0; col < cols; col++ {
		emptyRow := rows - 1
		for row := rows - 1; row >= 0; row-- {
			if grid[row][col] == 'a' {
				if row != emptyRow {
					grid[emptyRow][col] = 'a'
					grid[row][col] = '.'
				}
				emptyRow--
			} else if grid[row][col] == '#' {
				emptyRow = row - 1
			}
		}
	}
	return grid
}

// PrintGrid prints the grid
func PrintGrid(grid [][]rune) {
	for _, row := range grid {
		fmt.Println(string(row))
	}
}

func main() {
	filename := "apples.in"

	// Read input
	grid, rows, cols, err := ReadInput(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Simulate falling apples
	grid = SimulateFallingApples(grid, rows, cols)

	// Print the final grid
	PrintGrid(grid)
}

File: Falling Apples/apples.pl
use strict;
use warnings;

sub read_input {
    my ($file) = @_;
    open my $fh, '<', $file or die "Cannot open file: $!";
    my ($rows, $cols) = split ' ', <$fh>;
    my @grid;
    for (1..$rows) {
        chomp(my $line = <$fh>);
        push @grid, [split //, $line];
    }
    close $fh;
    return (\@grid, $rows, $cols);
}

sub simulate_falling_apples {
    my ($grid, $rows, $cols) = @_;
    for my $col (0..$cols-1) {
        my $empty_row = $rows - 1;
        for my $row (reverse 0..$rows-1) {
            if ($grid->[$row][$col] eq 'a') {
                if ($row != $empty_row) {
                    $grid->[$empty_row][$col] = 'a';
                    $grid->[$row][$col] = '.';
                }
                $empty_row--;
            } elsif ($grid->[$row][$col] eq '#') {
                $empty_row = $row - 1;
            }
        }
    }
    return $grid;
}

sub print_grid {
    my ($grid) = @_;
    for my $row (@$grid) {
        print join('', @$row), "\n";
    }
}

my $file = 'apples.in';
my ($grid, $rows, $cols) = read_input($file);
$grid = simulate_falling_apples($grid, $rows, $cols);
print_grid($grid);